geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloFeatureWriter.scala

```diff
@@ -71,6 +74,44 @@ object AccumuloFeatureWriter {
     val tablesAndNames = GeoMesaTable.getTablesAndNames(sft, ds)
     tablesAndNames.flatMap { case (table, name) => table.remover(sft).map((name, _)) }
   }
+
+  private val idGenerator: FeatureIdGenerator =
+    try {
+      logger.debug(s"Using feature id generator '${FEATURE_ID_GENERATOR.get}'")
+      Class.forName(FEATURE_ID_GENERATOR.get).newInstance().asInstanceOf[FeatureIdGenerator]
+    } catch {
+      case e: Throwable =>
+        logger.error(s"Could not load feature id generator class '${FEATURE_ID_GENERATOR.get}'", e)
+        new Z3FeatureIdGenerator
+    }
+
+  /**
+   * Sets the feature ID on the feature. If the user has requested a specific ID, that will be used,
+   * otherwise one will be generated. If possible, the original feature will be modified and returned.
+   */
+  def featureWithFid(sft: SimpleFeatureType, feature: SimpleFeature): SimpleFeature = {
+    if (feature.getUserData.containsKey(Hints.PROVIDED_FID)) {
+      withFid(sft, feature, feature.getUserData.get(Hints.PROVIDED_FID).toString)
+    } else if (feature.getUserData.containsKey(Hints.USE_PROVIDED_FID) &&
+        feature.getUserData.get(Hints.USE_PROVIDED_FID).asInstanceOf[Boolean]) {
+      feature
+    } else {
+      withFid(sft, feature, idGenerator.createId(sft, feature))
+    }
+  }
+
+  private def withFid(sft: SimpleFeatureType, feature: SimpleFeature, fid: String): SimpleFeature = {
+    try {
+      // FeatureIdImpl appears to be the only implementing class for FeatureId,
+      // but just in case wrap in a try/catch
+      feature.getIdentifier.asInstanceOf[FeatureIdImpl].setID(fid)
+      feature
+    } catch {
+      case e: ClassCastException =>
```

Comment:
yeah, i was trying to speed it up a little, and couldn't find any other implementations of FeatureId, so it seems like a real fringe case that it would be some other implementation - but I will change it.